core/libs/commonwealth/commonwealth/mavlink_comm/MavlinkComm.py

```diff
@@ -51,28 +68,58 @@ async def get_mavlink_message(
                 async with session.get(request_url, timeout=request_timeout) as response:
                     if not response.status == 200:
                         raise MavlinkMessageReceiveFail(f"Received status code of {response.status}.")
+                    # if message is "None", try re-detecting systemid
+                    if await response.text() == "None":
+                        self.set_system_id(await self.get_most_recent_vehicle_id())
+                        raise MavlinkMessageReceiveFail("Received empty response")
                     message = await response.json()
             except asyncio.exceptions.TimeoutError as error:
                 raise MavlinkMessageReceiveFail(f"Request timed out after {request_timeout} second.") from error
 
         return message
 
+    async def get_most_recent_vehicle_id(self) -> int:
+        json_data = await self.get_all_mavlink()
+        most_recent_timestamp = datetime.min.replace(tzinfo=timezone.utc)
+        most_recent_vehicle_id = None
+        for vehicle_id, vehicle in json_data["vehicles"].items():
+            # skip vehicles with id 2 (mavlink-router?) and 255 (common GCS)
+            if vehicle_id in {"2", "255"}:
+                continue
+
+            for component in vehicle["components"].values():
+                if "HEARTBEAT" in component["messages"]:
+                    message = component["messages"]["HEARTBEAT"]
+                    # we are looking for vehicles, not GCSs
+                    if not MavlinkVehicleType(message["mavtype"]["type"]).is_actually_a_vehicle():
+                        continue
+                    last_update_str = message["status"]["time"]["last_update"]
+                    last_update_str = re.sub(r"(\.\d{6})\d+", r"\1", last_update_str)
+                    last_update = datetime.fromisoformat(last_update_str)
+
+                    if last_update > most_recent_timestamp:
+                        most_recent_timestamp = last_update
+                        most_recent_vehicle_id = vehicle_id
+        return int(most_recent_vehicle_id or 1)
```

Comment:
```suggestion
        if most_recent_vehicle_id:
            logger.debug(f'{most_recent_vehicle_id = } (detected)')
            return int(most_recent_vehicle_id)
        else:
            logger.debug('no vehicle ID detected - using default (1)')
            return 1
```
Non-essential, but could come in handy at some point.